activerecord: Extract primary key constraints hash to a method

dylanahsmith opened the pull request:
cc @rafaelfranca 

Extract the primary key constraints hash, used for the WHERE scope to update or delete a record, into a method.  This is just a small step towards supporting a composite primary key, where we would want constraints on all primary key columns.

### Context

We have a use case for using a composite primary key for performance reasons, which isn't actually supported in rails at the moment.  Specifically, we use a parent record id along with the actual id to keep the associated records together for efficient querying (e.g. `(parent_id, id)`).

We can use `self.primary_key = :id` in the model to make most things work, along with a secondary `(id)` index.  However, that will end up taking gap locks in the innodb mysql engine when there is only an equality constraint on `id`, causing unnecessary lock contention.

jarthorn reviewed (commented):
There is one more reference in `update_columns` that can use the same hash: https://github.com/rails/rails/blob/main/activerecord/lib/active_record/persistence.rb#L697

dylanahsmith committed:
Extract primary key constraints hash to a method.

Since this hash will need to change in the future to support composite
primary keys.

dylanahsmith commented:
Changed `update_columns` to also use the extracted method.  Thanks
